### ISSUE TITLE
Core API: messages return inconsistent set with same start value

### ISSUE THREAD

**Issue by mkwong04** [zaproxy/zaproxy#5493]
I encounter inconsistent return results using the getHistory() methods.

The ZAProxy is monitoring traffic when i perform the getHistoryCount(), e.g. returned 30 which is assumed to be last position of time of taken.

after certain time i run getHistoryCount() again to obtain another value (e.g. 37).

So i use getHistory(30, 8) to retrieve the records between this 2 intervals, it return a set of 8 records

But after some time where traffic continue to flow, the same getHistory(30,8) return a different set of 8 records. 

Apparently there are some history record is "in progress" of persisting into the db and it alter the getHistory() return result (since it is ordered by historyId)

 Is there a consistent way to obtaining history set with same start and count value?

**Comment by psiinon** [comment]
Please dont raise issues issues for questions.
The ZAP User Group is a much better place for them: https://groups.google.com/group/zaproxy-users

**Comment by mkwong04** [comment]
Let me re-phrase it into issues

We are using the API in some automation

Scenario psuedo code
------------------------
//1. Invoke webdriver to load the webpage via ZAProxy 

//2. wait for 5 sec, find the last position in history
int marker = clientApi.core.**numberOfMessages**("")

//3. retrieve the reference history record in HAR format
HarLog markerHarLog = HarFileReaderUtil.readHarFile(clientApi.core.**messagesHar**(marker, 1));

//4. do some processing activities

//5. invoke webdriver to load another webpage 

//6. try to retrieve the same record using the marker
markerHarLog = HarFileReaderUtil.readHarFile(clientApi.core.**messagesHar**(marker, 1));

Issue
-----
at step 3, we obtain the records with **marker** position as start and count = 1, the records belongs _ZapMessageId = 31 (for example)

but at step 6, same value return records belongs to _ZapMessageId = 28 

Same start value to core.messagesHar() appear to return different record.

**Comment by thc202** [comment]
Did you check the type of the messages retrieved? Some messages created by ZAP are temporary and thus removed when no longer needed, that could explain the behaviour you are describing.

**Comment by mkwong04** [comment]
Yes, we only look for _ZapMessageType = 1 (Proxied message), actually we use the messagesHar to get blocked on histories and only extracted out those belongs to Proxied message type (originated from webdriver call)

If the temporary message created by ZAP is part of history and removed afterward AND assume the historyId of message is created in order, i should get a larger _ZapMessageId but i am getting a smaller value

Assume the following scenario
_ZapMessageId from 1 to 28 all proxied type message
_ZapMessageId = 29 (temporary Zap message)
_ZapMessageId = 30 (temporary Zap message)
_ZapMessageId = 31 <-- the last entry of history when the **numberOfMessage**() call that give count = 31

**messagesHar**(31,1) will give _ZapMessageId = 31

Now if temporary message removed.
_ZapMessageId from 1 to 28 all proxied type message
_ZapMessageId = 31 <- now position become 29

**messagesHar**(31,1) should give _ZapMessageId > 31

We consider this as an issue since the core.messagesHar() method's **start** parameter seems not working as it is expected (same value should return same start position record)

Could it be some messages is not persisted when the **numberOfMessage**() call?
